include/jemalloc/internal/arena_inlines_b.h

```diff
@@ -148,23 +148,23 @@ arena_decay_tick(tsdn_t *tsdn, arena_t *arena) {
 
 JEMALLOC_ALWAYS_INLINE void *
 arena_malloc(tsdn_t *tsdn, arena_t *arena, size_t size, szind_t ind, bool zero,
-    tcache_t *tcache, bool slow_path) {
+    bool slab, tcache_t *tcache, bool slow_path) {
 	assert(!tsdn_null(tsdn) || tcache == NULL);
 
 	if (likely(tcache != NULL)) {
-		if (likely(size <= SC_SMALL_MAXCLASS)) {
+		if (likely(size <= SC_SMALL_MAXCLASS && slab)) {
```

Comment:
I compiled locally with `-O3` and ThinLTO, and yes it appears it the compiler will reduce down to one branch when it can prove this is safe, like when we reach here from a call to `iallocztm`:


```asm
<jemalloc_je_iallocztm>:
;     bool is_internal, arena_t *arena, bool slow_path) {
               	push	rbp
               	push	r15
               	push	r14
               	push	r13
               	push	r12
               	push	rbx
               	sub	rsp, 0x1a8
               	mov	ebp, edx
               	mov	r13, rsi
               	mov	qword ptr [rsp + 0x10], rdi
               	mov	rsi, qword ptr [rsp + 0x1e0]
; 	if (likely(tcache != NULL)) {
               	test	r8, r8
               	je	 <L0>
               	mov	r12, r8
; 	return size <= SC_SMALL_MAXCLASS;
               	cmp	r13, 0x3800
; 		if (likely(size <= SC_SMALL_MAXCLASS && slab)) {
               	ja	 <L1>
; 	cache_bin_t *bin = &tcache->bins[binind];
               	mov	r15d, ebp
               	lea	r14, [r15 + 2*r15]
               	lea	rbx, [r12 + 8*r14]
               	add	rbx, 0x8
```